src/duration.rs

```diff
@@ -413,27 +413,30 @@ impl fmt::Display for Duration {
         // but we need to print it anyway.
         let (abs, sign) = if self.secs < 0 { (-*self, "-") } else { (*self, "") };
 
-        let days = abs.secs / SECS_PER_DAY;
-        let secs = abs.secs - days * SECS_PER_DAY;
-        let hasdate = days != 0;
-        let hastime = (secs != 0 || abs.nanos != 0) || !hasdate;
-
         write!(f, "{}P", sign)?;
-
-        if hasdate {
-            write!(f, "{}D", days)?;
+        // Plenty of ways to encode an empty string. `P0D` is short and not too strange.
+        if abs.secs == 0 && abs.nanos == 0 {
+            return f.write_str("0D");
         }
-        if hastime {
-            if abs.nanos == 0 {
-                write!(f, "T{}S", secs)?;
-            } else if abs.nanos % NANOS_PER_MILLI == 0 {
-                write!(f, "T{}.{:03}S", secs, abs.nanos / NANOS_PER_MILLI)?;
-            } else if abs.nanos % NANOS_PER_MICRO == 0 {
-                write!(f, "T{}.{:06}S", secs, abs.nanos / NANOS_PER_MICRO)?;
-            } else {
-                write!(f, "T{}.{:09}S", secs, abs.nanos)?;
+
+        f.write_fmt(format_args!("T{}", abs.secs))?;
+
+        if abs.nanos > 0 {
+            // Count the number of significant digits, while removing all trailing zero's.
+            let mut figures = 9usize;
+            let mut fraction_digits = abs.nanos;
+            loop {
```

Comment:
What's the point of this loop?

Reply:
To count the number of significant digits, and removing all trailing zero's. I.e. repeatedly divide `nanos` by 10.
I do not know a cleaner way to write it.